### PR TITLE
Implement Tuple.FlatMap

### LINKED ISSUE #13163: Missing scala.Tuple.flatMap
## Compiler version

3.0

## Description

We have `Tuple.FlatMap` to compute the type of `Tuple.flatMap` but we do not implement `Tuple.flatMap`.

We should implement it in a similar way as what is described in #12888.

### PR DESCRIPTION
**Pull request by KacperFKorban**
closes #13163

### CODE REVIEW

File: library/src/scala/Tuple.scala [hunk 1]
```diff
@@ -54,6 +54,12 @@ sealed trait Tuple extends Product {
   inline def map[F[_]](f: [t] => t => F[t]): Map[this.type, F] =
     runtime.Tuples.map(this, f).asInstanceOf[Map[this.type, F]]
 
+  /** Called on a tuple `(a1, ..., an)`, returns a new tuple `f(a1) ++ ... ++ f(an)`.
+   *  The result is typed as `Concat[F[A1], ...Concat[F[An], EmptyTuple]...])` if the tuple type is fully known.
+   */
+  inline def flatMap[F[_] <: Tuple](f: [t] => t => F[t]): FlatMap[this.type, F] =
```
Comment (nicolasstucki) [inline library/src/scala/Tuple.scala:60]:
```suggestion
  inline def flatMap[This >: this.type <: Tuple, F[_] <: Tuple](f: [t] => t => F[t]): FlatMap[This, F] =
```

**Comment by nicolasstucki** [comment]
You also need to update the `MiMaFilters`. You can run `scala3-interfaces/mimaReportBinaryIssues` to get the required list of `exclude`s.

File: tests/run/TupleFlatMap.scala [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  println(
+    Tuple().flatMap[EmptyTuple, [t] =>> (t, t)]([t] => (x: t) => (x, x))
+  )
+
+  println(
+    (1, 2, "x", "d").flatMap[(Int, Int, String, String), [t] =>> (t, t)]([t] => (x: t) => (x, x))
+  )
+
+  println(
+    (1, "x").flatMap[(Int, String), [t] =>> EmptyTuple]([t] => (x: t) => Tuple())
```
Comment (nicolasstucki) [inline tests/run/TupleFlatMap.scala:12]: The use of this API seems clunky. Do we need the explicit type parameters?

Comment (KacperFKorban) [reply tests/run/TupleFlatMap.scala:12]: I checked and it turns out that the explicit type params are needed. Though this PR shouldn't be merged yet.
Examples 3 and 4 result in a compilation error, but the first two examples end with compiler crash like:
```
Fatal compiler crash when compiling: tests/run/TupleFlatMap.scala:
assertion failed: orphan param: t, hash of binder = 1998366311, tree = <empty>, type = TypeVar(TypeParamRef(F) -> HKTypeLambda(List(T2), List(TypeBounds(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Nothing),TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Any))), AppliedType(TypeRef(ThisType(TypeRef(NoPrefix,module class scala)),class Tuple2),List(TypeParamRef(t), TypeParamRef(T2)))))
```
I'll take a look at it and try to fix it.

Comment (KacperFKorban) [reply tests/run/TupleFlatMap.scala:12]: Minimized it in #13286

File: library/src/scala/Tuple.scala [hunk 3]
```diff
@@ -54,6 +55,13 @@ sealed trait Tuple extends Product {
   inline def map[F[_]](f: [t] => t => F[t]): Map[this.type, F] =
     runtime.Tuples.map(this, f).asInstanceOf[Map[this.type, F]]
 
+  /** Called on a tuple `(a1, ..., an)`, returns a new tuple `f(a1) ++ ... ++ f(an)`.
+   *  The result is typed as `Concat[F[A1], ...Concat[F[An], EmptyTuple]...])` if the tuple type is fully known.
+   */
+  @experimental
+  inline def flatMap[This >: this.type <: Tuple, F[_] <: Tuple](f: [t] => t => F[t]): FlatMap[This, F] =
```
Comment (nicolasstucki) [inline library/src/scala/Tuple.scala:62]: This signature is not ideal as we need to pass the tuple type when we call. This should always be inferred.
Maybe we should wait until we can do something like
```scala
  inline def flatMap[F[_] <: Tuple][This >: this.type <: Tuple](f: [t] => t => F[t]): FlatMap[This, F] =
```

**Comment by nicolasstucki** [comment]
I will close this PR for now. We should revisit it in the future as the implementation is there but we still lack the ability to write the correct signature.

**Pull request closed without merging**